design/fast_finality.md

```diff
@@ -0,0 +1,295 @@
+a different light batch (e.g. one where the sequencer performs front-running) to be shared with peers. The enclave will 
+automatically reject a second light batch with the same parent.
+
+From the user's perspective, the transactions in the light batch are considered final (e.g. responses to RPC calls from
+the client behave as if the transactions were completely final).
+
+### Production of rollups
+
+Only the sequencer produces rollups, with the enclave code enforcing that only whitelisted enclaves can produce rollups.
+
+A rollup is produced whenever one of the following conditions is met:
+
+1. The number of transactions across all light-batches since the last rollup exceeds `x`
+2. The total size of all transactions across all light-batches since the last rollup exceeds `y`
+3. The number of blocks since the last rollup was produced exceeds `z`
+
+`x`, `y` and `z` are configurable per network. Rules (1) and (2) ensure that the rollup can fit within the Ethereum gas 
+limit. Rule (3) reduces the risk associated with the sequencer failing.
+
+To produce a rollup, the sequencer's host requests the creation of a rollup. The sequencer's enclave produces a rollup 
+containing all the light batches created since the last rollup, in a sparse Merkle tree structure. This rollup is 
+encrypted and signed, then sent to be included on the L1.
+
+The management contract on the L1 verifies that the rollup is produced by a designated sequencer.
+
+### Discovery of rollups
+
+Nodes scan incoming L1 blocks for new rollups. For each new rollup, the node checks that it contains all the light 
+batches produced since the last rollup. Each light batch contains the number of the rollup that will contain it. Since 
+the rollup is a sparse Merkle tree, proving non-inclusion of a given light batch is straightforward. If they discover 
+an issue, they can mount a challenge, as per the "Staking and slashing" section, below.
+
+The node then persists the rollup, so that they have a record of which light batches have been confirmed on the L1.
+
+### Staking and slashing
+
+The sequencer must put up a stake in the management contract. The management contract enforces that the sequencer must 
+keep their stake topped up to a suitable level to continue posting rollups.
+
+If a node finds that the contents of a given light-batch do not match the rollup published on the L1 (e.g. transactions 
+missing, transactions in the wrong order), it can post a challenge including the light batch and offending rollup to 
+the L1. The management contract will inspect this challenge. If successful, the sequencer will be slashed, with part of 
+their stake transferred to the challenger as a reward. This reward is greater than the cost of posting the challenge, 
+to incentivise prompt discovery of issues.
+
+Initially, there will be no staking mechanism to ensure light batches are produced at the correct cadence. Since there 
```

Comment:
I'd say "no slashing mechanism" rather than "staking mechanism" probably, it's the same stake as for the validation stuff.

Reply:
Thanks, fixed.